### PR TITLE
nextcloud30: init at 30.0.0; nextcloud30Packages: init

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

Changelog https://nextcloud.com/blog/nextcloud-hub9/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/nextcloud/default.nix [hunk 1]
```diff
@@ -1,6 +1,11 @@
-{ lib, stdenvNoCC, fetchurl, nixosTests
+{
+  lib
+, stdenvNoCC
+, fetchurl
+, nixosTests
```
Comment (dotlambda) [inline pkgs/servers/nextcloud/default.nix:5]: Why not follow nixfmt-rfc-style if we're refactoring?

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 2]
```diff
@@ -486,6 +486,12 @@
 
 - The kubelet configuration file can now be amended with arbitrary additional content using the `services.kubernetes.kubelet.extraConfig` option.
 
+- The latest available version of Nextcloud is v30 (available as `pkgs.nextcloud30`). The installation logic is as follows:
+  - If [`services.nextcloud.package`](#opt-services.nextcloud.package) is specified explicitly, this package will be installed (**recommended**)
+  - If [`system.stateVersion`](#opt-system.stateVersion) is >=24.11, `pkgs.nextcloud30` will be installed by default.
+  - If [`system.stateVersion`](#opt-system.stateVersion) is >=24.05, `pkgs.nextcloud29` will be installed by default.
```
Comment (dotlambda) [inline nixos/doc/manual/release-notes/rl-2411.section.md:492]:
```suggestion
  - If [`system.stateVersion`](#opt-system.stateVersion) is 24.05, `pkgs.nextcloud29` will be installed by default.
```
There is nothing in between.

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 3]
```diff
@@ -861,8 +861,8 @@ in {
               nextcloud defined in an overlay, please set `services.nextcloud.package` to
               `pkgs.nextcloud`.
             ''
-          else if versionOlder stateVersion "24.05" then nextcloud27
-          else nextcloud29
+          else if versionOlder stateVersion "24.11" then nextcloud29
+          else nextcloud30
         );
```
Comment (dotlambda) [inline nixos/modules/services/web-apps/nextcloud.nix:866]: Has 28 never been the default?

File: pkgs/servers/nextcloud/default.nix [hunk 4]
```diff
@@ -1,46 +1,60 @@
...
-      tests = nixosTests.nextcloud;
-      inherit packages;
-    };
+      src = fetchurl {
+        url = "https://download.nextcloud.com/server/releases/${pname}-${version}.tar.bz2";
+        inherit hash;
+      };
+
+      passthru = {
+        tests = nixosTests.nextcloud;
```
Comment (dotlambda) [inline pkgs/servers/nextcloud/default.nix:30]: It would be nice to have this only be the tests with the appropriate major version suffix, but I can make that change in a separate PR.

File: pkgs/servers/nextcloud/default.nix [hunk 5]
```diff
@@ -1,46 +1,60 @@
...
+    stdenvNoCC.mkDerivation rec {
+      pname = "nextcloud";
+      inherit version;
 
-    passthru = {
-      tests = nixosTests.nextcloud;
-      inherit packages;
-    };
+      src = fetchurl {
+        url = "https://download.nextcloud.com/server/releases/${pname}-${version}.tar.bz2";
```
Comment (dotlambda) [inline pkgs/servers/nextcloud/default.nix:25]:
```suggestion
        url = "https://download.nextcloud.com/server/releases/nextcloud-${version}.tar.bz2";
```
#277994

File: pkgs/servers/nextcloud/default.nix (hunk 4 above)
Comment (dotlambda) [reply pkgs/servers/nextcloud/default.nix:30]: https://github.com/NixOS/nixpkgs/pull/342896

File: pkgs/servers/nextcloud/default.nix [hunk 6]
```diff
@@ -1,46 +1,60 @@
...
+        homepage = "https://nextcloud.com";
+        maintainers = with lib.maintainers; [
+          schneefux
+          bachp
+          globin
+          ma27
+        ];
+        license = lib.licenses.agpl3Plus;
+        platforms = lib.platforms.linux;
+        knownVulnerabilities = lib.extraVulnerabilities ++ (lib.optional eol "Nextcloud version ${version} is EOL");
```
Comment (dotlambda) [inline pkgs/servers/nextcloud/default.nix:53]:
```suggestion
        knownVulnerabilities = extraVulnerabilities ++ (lib.optional eol "Nextcloud version ${version} is EOL");
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by onny** [comment]
Thank you @dotlambda fixed the issues

File: pkgs/servers/nextcloud/default.nix [hunk 7]
```diff
@@ -1,48 +1,63 @@
...
-, nextcloud29Packages
+{
+  lib,
+  stdenvNoCC,
+  fetchurl,
+  nixosTests,
+  nextcloud28Packages,
+  nextcloud29Packages,
+  nextcloud30Packages,
 }:
```
Comment (Ma27) [inline pkgs/servers/nextcloud/default.nix:9]: Please split this into another commit. Reviewing a mixture of behavioral and stylistic changes is always annoying.

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 8]
```diff
@@ -862,7 +862,8 @@ in {
               `pkgs.nextcloud`.
             ''
           else if versionOlder stateVersion "24.05" then nextcloud27
-          else nextcloud29
+          else if versionOlder stateVersion "24.11" then nextcloud28
+          else nextcloud30
```
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:866]: I think we don't need the `else if` for `24.11` here, just do `else nextcloud30`: if you have a fresh NixOS install (i.e. stateVersion = 24.11) you ended up on nextcloud29 originally, so going straight to nextcloud30 is still safe since we move forward one major-release only.

Or am I missing something?

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 9]
```diff
@@ -507,6 +507,12 @@
 
 - The kubelet configuration file can now be amended with arbitrary additional content using the `services.kubernetes.kubelet.extraConfig` option.
 
+- The latest available version of Nextcloud is v30 (available as `pkgs.nextcloud30`). The installation logic is as follows:
+  - If [`services.nextcloud.package`](#opt-services.nextcloud.package) is specified explicitly, this package will be installed (**recommended**)
+  - If [`system.stateVersion`](#opt-system.stateVersion) is >=24.05, `pkgs.nextcloud29` will be installed by default.
+  - If [`system.stateVersion`](#opt-system.stateVersion) is >=24.11, `pkgs.nextcloud30` will be installed by default.
```
Comment (Ma27) [inline nixos/doc/manual/release-notes/rl-2411.section.md:513]: This will need an update with the comment above.

**Review by Ma27 — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 8 above)
Comment (bachp) [reply nixos/modules/services/web-apps/nextcloud.nix:866]: I think anyway the `netcloud28` would be wrong as  it would mean a downgrade for current 24.11 users.

But I agree with @Ma27 that we should just go to `nextcloud30` and no if needed

Comment (dotlambda) [reply nixos/modules/services/web-apps/nextcloud.nix:866]: Maybe we can make things simpler going forward if we don't specify a default package at all.

Comment (onny) [reply nixos/modules/services/web-apps/nextcloud.nix:865]: I would like to keep the default package version, it will help automatically upgrade major Nextcloud version on new state versions

**Comment by onny** [comment]
@Ma27 hopefully addressed the remaining issues :)

**Review by dotlambda — APPROVED** [review]

**Comment by Ma27** [comment]
WIll take another look today.

**Review by helsinki-Jo — CHANGES_REQUESTED** [review]

**Comment by Ma27** [comment]
> WIll take another look today.

Sorry, will take me until tomorrow.

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 10]
```diff
@@ -507,6 +507,12 @@
 
 - The kubelet configuration file can now be amended with arbitrary additional content using the `services.kubernetes.kubelet.extraConfig` option.
 
+- The latest available version of Nextcloud is v30 (available as `pkgs.nextcloud30`). The installation logic is as follows:
+  - If [`services.nextcloud.package`](#opt-services.nextcloud.package) is specified explicitly, this package will be installed (**recommended**)
+  - If [`system.stateVersion`](#opt-system.stateVersion) is >=24.05, `pkgs.nextcloud28` will be installed by default.
```
Comment (Ma27) [inline nixos/doc/manual/release-notes/rl-2411.section.md:512]: If we keep the stateVersion part, it's `nextcloud29` here.

File: pkgs/servers/nextcloud/packages/28.json
[the remark this answers is not in the document]
Comment (Ma27) [reply pkgs/servers/nextcloud/packages/28.json:1]: May be nice to have it in another commit, but I don't think it's worth splitting this into two PRs :shrug:

[the remark this answers is not in the document]
Comment (Ma27) [reply pkgs/servers/nextcloud/packages/28.json:1]: Performed the split, retained the original author information while doing so.

**Review by Ma27 — APPROVED** [review]
Current state is working fine for me (except I'm unhappy that `maps` isn't ready for nc30 yet :cry: ).
